### DISCUSSION TITLE
connect() ??

### DISCUSSION THREAD

**Discussion by alabamatoy** [jgyates/genmon#1093] [Questions]
### Checklist

- [X] I made sure that there are *no existing issues or discussions* - [open issues](https://github.com/jgyates/genmon/issues),[closed issues](https://github.com/jgyates/genmon/issues?q=is%3Aissue+is%3Aclosed),[open discussion](https://github.com/jgyates/genmon/discussions), [closed discussion](https://github.com/jgyates/genmon/discussions/categories/questions?discussions_q=is%3Aclosed) - which may answer my question or that I could contribute my information to.
- [X] I have read the [Known Issues Wiki Page](https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues) and my question / issue is not listed.
- [X] I have provided the version of the software and the operating system in my post.
- [X] If I have an issue with genmon not communicating with the controller, I have read and followed the [serial troubleshooting guide](https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting).

### Version

1.19.02

### Have you submitted your log files from the About page? Note that outbound email must be working for logs to be submitted.

I have submitted logs

### What is/are your question(s)?

I get the following upon clicking on the test email:

```
An error occurred:
Error sending test email :
Please run connect() first

Please try again
```

In genmon log I see:
`2024-03-28 13:12:19,121 : Error sending test email : please run connect() first : genserv.py:726`

Im unable to determine how to "run connect()".  Im pretty sure the SMTP settings are correct, per the provider's website: "(Outgoing Mail): Server: smtp.epbfi.com, SSL: Yes, Port: 465."  In settings I have SSL on and disable TLS on.

### Additional information

I have thunderbird on a machine on same subnet with these settings and its able to send and receive email.

**Comment by jgyates** [comment]
The discussion and issues threads have tags that can be searched by subject. The "email" tag search will show how issues like this have been resolved. SSL does work, TLS is a more recent and more secure method. Attach a screen shot of your setting if the past issues do not resolve your issue.

**Comment by jgyates** [comment]
Here is a thread from 2 days ago that sounds very similar to your issue:  https://github.com/jgyates/genmon/discussions/1091

The link includes both screen shots of working system and the solution for someone who could not get SSL to work.

**Reply by alabamatoy** [reply]
I had read through #1091 while wrestling with this.  I may be daft here, but I just dont get it.  Here's my settings:
![emailsettings](https://github.com/jgyates/genmon/assets/28680526/6e9945ad-fd02-4c11-ab85-74031f4ce7b0)
The service support page sez: 
![epbsupport](https://github.com/jgyates/genmon/assets/28680526/00887820-31c8-4aeb-91e6-18a140ea4be2)
I know its talking to the server because if I mangle the PW, it errors out with an authentication error.  The error message about running connect() is really not helpful.  Would be nice if the logs would capture what the actual error is.

**Reply by jgyates** [reply]
Your settings look OK for SSL. The "Disable SMTP Auth" is server dependent so I don't know what your SMTP server requires. You can try it with and without. Most servers require SMTP authentication so if it does not make a difference in your testing, leave "Disable SMTP Auto" off.

If you save your settings on the settings page you can look at the file /var/log/mymail.log for mail specific errors. Also, you can manually edit the file /etc/genmon/mymail.conf and add the following line:

     debug=True

Save the file and restart genmon. With this enabled you will get more debugging output in /var/log/mymail.log. Note that you should not leave the debug=True enabled after you resolve the issue as this will, over time, decrease the life of your SD card as it increases the writes to the log file (unless you are logging to a RAM disk as recommended in the wiki).

Can you ping smtp.epbfi.com from your pi? If you have a static IP and have not configured /etc/resolve.conf with a DNS server iP address then you will not be able to resolve the domain name to an IP.

The other suggestions are:

1) Try using TLS and port 587 instead of SSL on port 465. This is more secure, newer standard and the recommended protocol for sending mail
2) Try setting "Disable TLS Encryption" to off
3) Make sure you are using an email account that you can either bypassing Multi factor authentication (preferred) or disable multi factor authentication. For example gmail requires multi factor authentication to be enabled on all accounts as of 5/2022, however you can create application specific passwords for use with sending or receiving email. This is the method used for configuring gmail with genmon. See the [wiki page here](https://github.com/jgyates/genmon/wiki/Appendix-M-Tips-and-Tricks-useful-for-Genmon#setup-gmail-for-use-with-genmon).
4) Use a gmail, outlook, hotmail, or another free email provider. Unfortunately I can only validate with email server that I have an account on however based on the log submissions over the years there are a lot of different domains that send email via genmon (have not examined the headers to see the originating email server). Also, keep in mind that this account is just for sending notices to you if someting is wrong with the generator, or to the developer (me) if something is unexplained or unexpected in the data back from the controller so the sending email account domain not really important. 

The SNMP support in this project is handled via the [snmtplib python library](https://docs.python.org/3/library/smtplib.html). The genmon implantation of this code is [here](https://github.com/jgyates/genmon/blob/master/genmonlib/mymail.py). While there have been minor changes to the mail code in the past couple of years, it has been for to add logging or improve error handing. The base code for authenticating and sending has not changed in years.

**Comment by alabamatoy** [comment]
From /var/log/mymail.log:
```
2024-03-29 17:21:10,471 : Error SMTP sendmail: (554, b'5.7.1 [VO-1] Message blocked due to spam content in the message.') : mymail.py:666
2024-03-29 17:21:10,473 : Error in SendMailThread, retrying (2): please run connect() first : mymail.py:692

```
Respectfully, this whole prolonged troubleshooting goat-dance by the dullard on this end could have been alleviated if either the error message was captured from the SMTP server response and displayed to me or it instead told me to check /var/log/mymail.log for further info.  The "run connect() first" is not only misleading, but in this case its flat wrong.  I mean this in the most constructive manner possible....

Now, as to the spam content, I have no idea what the SMTP server is finding distasteful, because I dont know what your test message looks like.  I  probably should call the ISP's helpdesk and try to get it unblocked......they have been very helpful to me in the past once I get past the first tier support.  But if this server is finding it unsavory, chances are others will also.

**Reply by jgyates** [reply]
@alabamatoy 

I agree that it would be ideal if the original error message was more useful. I all I do is change things as I see them. This particular issues is new as I have never seen this error before. In the past most errors are on the authentication steps, not the actual sending. This issues appears to be that you got two errors and the UI only displays the last one. The 2nd error is likely a by product of the first (my best guess based on the info provided and looking at the code).

I checked in an update that I hope fixes the issue with the error display. I am not 100% sure that it will as I don't have the ability to test this particular failure so let me know if the display behavior changes once you update from the about page (version number will not change). The code change should now display the original error if a failure on send occurs. Email related errors at run time are still logged to /var/log/mymail.log.

Regarding the messages, sent from genmon. All messages are sent in MIME Base 64 text encoding which is a very simple type of message. It is possible that they are complaining about the frequency or the sending email address. Maintaining an email server of any kind is a constant battle that requires a lot of work to fend off spammers and hackers, which can lead to more resource constrained ISPs leaning toward security vs openness / usability. For IoT type applications this is one reason I don't use my ISPs email services, but opt to use gmail.

There are 3 messages in total: 

1) this is the message sent when you send a test email (any failures are not logged to /var/log/mymail.log but the last error is displayed):
     
Subject: Genmon Test Email

        Time: 06:25:19 PM
        Date: Fri 29-Mar-2024
        Test email from genmon

2) This message is sent when genmon starts (any errors are logged to /var/log/mymail.log):
 
Subject: Generator Monitor Starting at SiteName

        Time: 06:15:01 PM 
        Date: Fri 29-Mar-2024
        Generator Monitor Starting at SiteName using IP address 192.168.1.10
      
The IP address is the IP address assigned to the Pi

3) This message is sent when genmon starts and after the generator state has been established (any errors are logged to /var/log/mymail.log):

Subject: Generator Notice: SiteName
        
        Time: 06:15:14 PM
        Date: Fri 29-Mar-2024
        
        For additional information : http://192.168.11.14:8000/
        NOTE: This message is a notice that the state of the generator has changed. The system is not in alarm.

Then the contents of the Status page text (no gauges) is included in the message body

**Comment by alabamatoy** [comment]
![image](https://github.com/jgyates/genmon/assets/28680526/707df515-06c6-43c5-992d-06913a7d2057)

TYVM

**Reply by skipfire** [reply]
Have you added a sender name and address in your config?

**Reply by alabamatoy** [reply]
The address field help says if its blank it uses the auth account email, which should work.  I added a name, but it still fails.  Added an address, still fails.  (Yes, I saved settings in between each change) I think its a configuration setting on the SMTP server.  The ISP is looking into it.

**Reply by alabamatoy** [reply]
BTW, I have another server on this same subnet using this same SMTP server for logcheck, and it works, just sent me an automated/timed  message from that server.  So its something unique to the Genmon setup/message coms that is causing the SMTP server to refuse the message.

**Reply by jgyates** [reply]
OK. Let me know once you have more info from your ISP. I don't have any hint as what would cause this issue as this is the first time in 6 years of this project that I have seen this issue. Having said that I don't have an issue changing the format as long as it is not causing more issues than it is correcting.

**Reply by jgyates** [reply]
@alabamatoy 

I checked in an update that *might* help. I an not 100% sure this will change anything, but it is worth a shot to test.

First update from the about page, once genmon restarts, see if you get any emails from genmon when it restarts. Then try to send a test email on the Settings page.

If that works then great. If it does not then I have one more thing to try:

Edit the file /etc/genmon/mymail.conf file and add this line to the bottom of the file and save the file.

     use_html=True

Then restart genmon. Check to see if you receive any emails when genmon starts. Note that the test email on the settings page will not (yet) have the option of sending and email formatted as HTML instead of text. I will not add this option in the user interface until I get some feedback from you with your testing.

If you get failures, post the output of the error in the UI and in /var/log/mymail.log

Let me know how it goes.

**Reply by jgyates** [reply]
I am going to close this issue but let me know if you have any status update or questions.

**Reply by jgyates** [reply]
@alabamatoy, Were you able to resolve this issue?

**Reply by alabamatoy** [reply]
I have given up.  There is something in the way that genmon talks to my ISP SMTP server that the server doesnt like.  The ISP cant or wont fix it.  The truly strange thing is, I have another server, on the same subnet (from the outside its same IP!) which talks to my ISP and I can relay genmon through that successfully  I dont know what the problem is, but sending through my internal postfix relay is working, and I'm going to move on.  
Thanks for all your help.

**Reply by jgyates** [reply]
ok. If you ever want to pick it up again you can just post to this thread. I am glad you found a work around.

**Comment by alabamatoy** [comment]
I've been focused elsewhere and traveling, will be back l8r this week and
will address further.



Herndon Elliott
